src/main/java/com/vlkan/pubsub/PubsubSubscriber.java

```diff
@@ -0,0 +1,222 @@
+/*
+ * Copyright 2019 Volkan Yazıcı
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permits and
+ * limitations under the License.
+ */
+
+package com.vlkan.pubsub;
+
+import com.vlkan.pubsub.model.PubsubAckRequest;
+import com.vlkan.pubsub.model.PubsubPullRequest;
+import com.vlkan.pubsub.model.PubsubPullResponse;
+import com.vlkan.pubsub.model.PubsubReceivedAckableMessage;
+import com.vlkan.pubsub.ratelimiter.StagedRateLimiter;
+import com.vlkan.pubsub.util.FluxHelpers;
+import io.micrometer.core.instrument.DistributionSummary;
+import io.micrometer.core.instrument.MeterRegistry;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import reactor.core.publisher.Flux;
+import reactor.core.publisher.Mono;
+import reactor.core.scheduler.Scheduler;
+
+import java.math.BigDecimal;
+import java.time.Duration;
+import java.util.List;
+import java.util.Objects;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+public class PubsubSubscriber {
```

Comment:
Same as in my comment on `PubsubPublisher`. `Subscriber` suffix here is very misleading

Reply:
Let's continue this discussion in the `PubsubPublisher` comments.